external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -245,44 +291,221 @@ impl FieldInfo {
+impl std::fmt::Display for ResolvedModuleMember {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            ModuleAccessKind::Function => write!(f, "function"),
-            ModuleAccessKind::Constant => write!(f, "constant"),
-            ModuleAccessKind::Datatype => write!(f, "type"),
+            ResolvedModuleMember::Function(_) => write!(f, "function"),
+            ResolvedModuleMember::Constant(_) => write!(f, "constant"),
+            ResolvedModuleMember::Datatype(ty) => write!(f, "{}", ty.datatype_kind_str()),
+        }
+    }
+}
+
+//**************************************************************************************************
+// Module Index
+//**************************************************************************************************
+// This indes is used for looking full paths up in name resolution.
+
+pub type ModuleMembers = BTreeMap<ModuleIdent, BTreeMap<Symbol, ResolvedModuleMember>>;
+
+pub fn build_member_map(
+    env: &CompilationEnv,
+    pre_compiled_lib: Option<Arc<FullyCompiledProgram>>,
+    prog: &E::Program,
+) -> ModuleMembers {
+    // NB: This checks if the element is present, and doesn't replace it if so. This is congruent
+    // with how top-level definitions are handled for alias resolution, where a new definition will
+    // not overwrite the previous one.
+    macro_rules! add_or_error {
+        ($members:ident, $name:expr, $value:expr) => {{
+            let name = $name.value();
+            if $members.contains_key(&name) {
+                assert!(env.has_errors());
+            } else {
+                $members.insert(name, $value);
+            }
+        }};
+    }
+
+    use ResolvedModuleMember as M;
+    let all_modules = prog
+        .modules
+        .key_cloned_iter()
+        .chain(pre_compiled_lib.iter().flat_map(|pre_compiled| {
+            pre_compiled
+                .expansion
+                .modules
+                .key_cloned_iter()
+                .filter(|(mident, _m)| !prog.modules.contains_key(mident))
+        }));
+    let mut all_members = BTreeMap::new();
+    for (mident, mdef) in all_modules {
+        let mut members = BTreeMap::new();
+        for (name, sdef) in mdef.structs.key_cloned_iter() {
+            let tyarg_arity = sdef.type_parameters.len();
+            let field_info = match &sdef.fields {
+                E::StructFields::Positional(fields) => FieldInfo::Positional(fields.len()),
+                E::StructFields::Named(f) => {
+                    FieldInfo::Named(f.key_cloned_iter().map(|(k, _)| k).collect())
+                }
+                E::StructFields::Native(_) => FieldInfo::Empty,
+            };
+            let struct_def = ResolvedStruct {
+                mident,
+                name,
+                decl_loc: name.loc(),
+                tyarg_arity,
+                field_info,
+            };
+            add_or_error!(
```

Comment:
I'm confused, when would this have a duplicate?

Reply:
Ah, good point. For structs, they already must be unique.

For the others, tthis code is dumping all module members into the same name map (since we already produce errors when they overlap) for uniform resolution with better error reporting. We will now sometimes produce an error when `A` is both a function and a constant and you try to use it as both ("the term `A` resolves to a function") but this seems sane to me. Variables still live in a separate namespace, of course.